functions/src/update-stats.ts

```diff
@@ -24,47 +20,50 @@ export const updateStats = functions.firestore
       return;
     }
     const vote = voteSnapshot.data()! as Vote;
-    let stats: StatsData;
+    const voteId: string = voteSnapshot.id
     await firebase.firestore().runTransaction(async transaction => {
-      await transaction
-        .get(
+      const doc = await transaction.get(
+        firebase
+          .firestore()
+          .collection("stats")
+          .doc(vote.campaign)
+      );
+      if (doc.get(voteId)) {
+        console.log("this vote has already been counted");
+        throw new functions.https.HttpsError(
+          "already-exists",
+          `This vote has already been counted`
+        );
+      }
+
+      let data = doc.data();
+      if (!data) {
+        data = {};
+        data[vote.value] = 1;
+      } else {
+        if (!data[vote.value]) {
+          data[vote.value] = 1;
+        } else {
+          data[vote.value] += 1;
+        }
+      }
+
+      return transaction
+        .set(
           firebase
             .firestore()
             .collection("stats")
-            .doc(vote.campaign)
+            .doc(vote.campaign),
+          data
         )
-        .then(doc => {
-          stats = doc.data()! as StatsData;
-          if (!stats) {
-            stats = {
-              greatCount: 0,
-              notThatGreatCount: 0,
-              notGreatAtAllCount: 0
-            };
-          }
-          switch (vote.value) {
-            case validVotes[0]:
-              stats.greatCount += 1;
-              break;
-            case validVotes[1]:
-              stats.notThatGreatCount += 1;
-              break;
-            case validVotes[2]:
-              stats.notGreatAtAllCount += 1;
-              break;
-          }
-
-          return transaction.set(
-            firebase
-              .firestore()
-              .collection("stats")
-              .doc(vote.campaign),
-            {
-              greatCount: stats.greatCount,
-              notThatGreatCount: stats.notThatGreatCount,
-              notGreatAtAllCount: stats.notGreatAtAllCount
-            }
-          );
-        });
+        .set(
+          firebase
+            .firestore()
+            .collection("stats")
+            .doc(vote.campaign)
```

Comment:
`firebase.firestore().collection("stats").doc(vote.campaign)` is used 3 times in the function. You might want to store it in a variable to avoid mistakes.